PEP 734: Multiple Interpreters in the Stdlib

ericsnowcurrently opened the pull request:
<!--
You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

If your PEP is not Standards Track, remove the corresponding section.
-->

## Basic requirements (all PEP Types)

* [ ] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [ ] File created from the [latest PEP template](https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [x] Core dev/PEP editor listed as ``Author`` or ``Sponsor``, and formally confirmed their approval
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate
* [x] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required)
* [x] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
* [x] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [x] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale
    * [x] Specification
    * [ ] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [ ] Reference Implementation
    * [x] Rejected Ideas
    * [ ] Open Issues
* [x] ``Python-Version`` set to valid (pre-beta) future Python version, if relevant
* [ ] Any project stated in the PEP as supporting/endorsing/benefiting from the PEP formally confirmed such
* [ ] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``


<!-- readthedocs-preview pep-previews start -->
----
:books: Documentation preview :books:: https://pep-previews--3523.org.readthedocs.build/pep-0734/

<!-- readthedocs-preview pep-previews end -->

JelleZijlstra commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,399 @@
+how users will make use of multiple interpreters in Python code, we are
+purposefully keeping the initial API as lean and minimal as possible.
+The objective is to provide a well-considered foundation on which we may
+add more-advanced functionality later.
```
```suggestion
add more advanced functionality later.
```

ericsnowcurrently replied:
The hyphen was intentional, but your comment demonstrates that it may be awkward. 😄  I'll reword the phrase.

JelleZijlstra commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,399 @@
+
+Additional details:
+
+* Every ``Interpreter`` instance wraps an ``InterpreterID`` object.
```
Is this a public type? Is it the type of the `.id` attribute mentioned above?

ericsnowcurrently replied:
It is a public type and will be exposed as `interpreters.InterpreterID`.  However, I wasn't planning on including that in the PEP.  I'll reword the paragraph.

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+
+The module introduces a basic communication mechanism called "channels".
+They are based on `CSP`_, as is ``Go``'s concurrency model (loosely).
+Channels are like pipes: FIFO queues with distinct send/recv ends.
```
?
```suggestion
Channels are like pipes: FIFO queues with distinct send/receive ends.
```

ericsnowcurrently replied:
I was mirroring the type and method names.  However, that doesn't add much here.  I'll spell it out.

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+| ``.recv() -> object``                    | | Get the next object from the channel,       |
+|                                          | | and wait if none have been sent.            |
++------------------------------------------+-----------------------------------------------+
+| ``.recv_nowait(default=None) -> object`` | | Like recv(), but return the default         |
```
```suggestion
| ``.recv_nowait(default=None) -> object`` | | Like ``recv()``, but return the default     |
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+| ``.send(obj)``               | | Send the `shareable object <Shareable Objects_>`_ (i.e. its data) |
+|                              | | to the receiving end of the channel and wait.                     |
++------------------------------+---------------------------------------------------------------------+
+| ``.send_nowait(obj)``        | Like send(), but return False if not received.                      |
```
```suggestion
| ``.send_nowait(obj)``        | Like ``send()``, but return ``False`` if not received.              |
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+
+::
+
+   tasks_recv, tasks = interpreters.create_channel()
```
```suggestion
   import interpreters

   tasks_recv, tasks = interpreters.create_channel()
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+   for t in threads:
+       t.join()
+
+Sharing a memoryview (imagine map-reduce):
```
```suggestion
Sharing a ``memoryview`` (imagine map-reduce):
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+Rejected Ideas
+==============
+
+See :pep:`554`.
```
```suggestion
See :pep:`PEP 554 <554#rejected-ideas>`.
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+* much of the API information in this PEP
+
+Docs about resources for extension maintainers already exist on the
+`Isolating Extension Modules <isolation-howto_>`_ howto page.  Any
+extra help will be added there.  For example, it may prove helpful
+to discuss strategies for dealing with linked libraries that keep
+their own subinterpreter-incompatible global state.
+
+.. _isolation-howto:
+   https://docs.python.org/3/howto/isolating-extensions.html
```
We can link using Intersphinx:
```suggestion
:ref:`python:isolating-extensions-howto` howto page.  Any
extra help will be added there.  For example, it may prove helpful
to discuss strategies for dealing with linked libraries that keep
their own subinterpreter-incompatible global state.
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+more easily accessible to Python code.  This is particularly relevant
+now that we have a per-interpreter GIL (:pep:`684`) and people are
+more interested in using multiple interpreters.  Without a stdlib
+module, users are limited to the `C-API`_, which restricts how much
+they can try out and take advantage of multiple interpreters.
+
+.. _C-API:
+   https://docs.python.org/3/c-api/init.html#sub-interpreter-support
```
Using Intersphinx:
```suggestion
module, users are limited to the :ref:`C API <python:sub-interpreter-support>`, which restricts how much
they can try out and take advantage of multiple interpreters.
```

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,400 @@
+  finalized.  Thus no interpreters created through
+  ``interpreters.create()`` will leak.
+
+|
+
```
No.  In my editor the bullets felt a bit cramped.  However, in the preview I see it shouldn't be a problem.

gvanrossum reviewed (commented):
Hi Eric, I know we already spoke about some of this off-line, but I ran out of time reviewing all my comments. I know the next revision will be much improved!

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+of the target interpreter.  This is the same as how it works
+for code run from the Python command-line (e.g. ``-m``) or the REPL.
+
+The builtin ``exec()`` discards any object returned from the
+executed code.
```
I agree, there is no way to return an object from `exec()`.  Any return value is discarded.  Normally what one passes to `exec()` wouldn't have a `return` statement.  However, it actually can:

```py
def func():
    global spam
    spam = True
    return 'a returned value'

ns = {}
res = exec(func.__code__, ns)
# Nothing is returned.  The returned string was discarded.
assert res is None, res
assert ns['spam'] is True, ns
```

gvanrossum replied:
Oh, I had forgotten about that. It might help if you called out this edge case. :-)

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+(in the same way as the ``threading`` module).  However, that low-level
+API is not intended for public use and thus not part of this proposal.
+
+We also expect that an ``InterpreterPoolExecutor`` will be added to the
+``concurrent.futures`` module, but that is outside the scope of this PEP.
```
I don't mind pulling it in.

gvanrossum replied:
Great. In the previous PEP version, this didn't follow the Executor model very closely -- instead of a function and optional arguments you were passing a string (like `Interpreter.exec()`) IIUC `ProcessPoolExecutor` has solved this issue, so hopefully `InterpreterPoolExecutor` can solve it similarly. (I'm guessing using `pickle`? Though I've never looked.)

ericsnowcurrently replied:
Yeah, it would be pickle.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
++----------------------------------------------------+-----------------------+
+| signature                                          | description           |
++====================================================+=======================+
+| ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |
```
`RecvChannel` is the receiving end and `SendChannel` is the sending end.  Perhaps it would help if they were something like `RecvChannelEnd` and `SendChannelEnd`?

gvanrossum replied:
Hm, those don't delight me, but I can't think of something better either. I wonder if something using sender/receiver or reader/writer could be used?

The key insight that I eventually reached was that this returns two ends of a queue, and the API is a bit like `r, w = os.pipe()`. Including the part where you need to "give" the other party the other end of a channel you created. Which you can do using either `set_main_attr()` or by sending it over an already established channel. And that's why channels are "shareable" objects (rather uniquely, since all the other shareable types are basic data types).

ericsnowcurrently replied:
Honestly, I'm probably going to drop the whole "distinct ends" thing, have one class, and just call it `Queue`, in the spirit of the minimal API I'm aiming for.  For that matter, I suppose we could put it in the `queue` module (something like `CrossInterpreterQueue`), but it probably still belongs in `interpreters`.  We could add directional channels later, if worth doing.

My only reservation is that single-directional channel ends often makes code easier to reason about and harder to do the wrong thing.  A bi-directional queue object has the opposite effect.  At least, that was my experience with Go, which has both directional and bi-directional channels.  A bi-directional channel was never necessary but it was easy to use one without thinking (at the expense of extra cognitive overhead later), since that's what you get when you create a channel (e.g. `var foo = make(chan int)`).  Having only directional channels from the start eliminates the problem, which is part of why I've been proposing `RecvChannel` and `SendChannel`.  In some ways it's better to not introduce the less desirable thing in the first place, rather than try to discourage its use (or remove it) later.

However, given your reaction, I'm inclined to think it isn't worth having the distinct ends.  Furthermore, I don't recall anyone bringing up concerns with Python's `queue.Queue` like I've experienced with Go's bi-directional channel.  So it may very well be a Go-only issue and one I needn't worry about with the `interpreters` module.

gvanrossum replied:
A single Queue class sounds good, just don't put it in queue.py (it's not really related except at a high level. For example, asyncio has its own queue submodule.

I wouldn't worry too much about not having distinct read/write ends -- we live with that for file I/O too, for example. I presume that using a Queue, an interpreter could possibly send to itself, which seems fine as an edge case.

Is there something about queue capacity?

ericsnowcurrently replied:
> Is there something about queue capacity?

I didn't see a need to worry about anything but unbounded for now.

gvanrossum replied:
Okay, but then why do we have both `send()` and `send_nowait()`? If the queue is unbounded there's never a need to block in `send()`, is there? How is the difference implemented?

(I have no beef with `recv_nowait()` -- if the queue is empty, one blocks, the other doesn't. But in queue.py, `put_nowait()` exists to avoid blocking when the queue is *full*.)

ericsnowcurrently replied:
In that context you can think of the queue as having a max size of 0 then.

FWIW, I've found it useful to know when the object is actually popped off the queue, as a synchronization point between two interpreters.

gvanrossum replied:
Oh wait, the updated description makes it clear that queues have max capacity zero. That's surprising given that previously you described it as similar to `os.pipe()`, but it is a consistent model, so I withdraw my objections.

Now I vaguely wonder how we'd implement a queue *with* buffering on top of this, but that's a straightforward programming problem.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+
+The module introduces a basic communication mechanism called "channels".
+They are based on `CSP`_, as is Go's concurrency model (loosely).
+Channels are like pipes: FIFO queues with distinct send/receive ends.
```
I'm not quite sure what you mean by "1:1 correspondence".  Each sent object is added to the back of the channel's internal queue and each `recv()` pops the next one off the front of that queue.  Every sent object will be received.

gvanrossum replied:
Okay, sorry, yes, that's what I meant by 1:1 correspondence. For pipes this isn't the case, unless you implement some higher-level protocol: when using pipes, one end can write two lines using two `write()` calls, and the other end *may* receive them all in a single `read()` call (and data written using a single `write()` call may be spread across multiple `read()` calls).

Maybe it's better not to mention pipes and just state that channels are queues?

(Which reminds me -- channels are thread-safe, right?)

ericsnowcurrently replied:
Yeah, channels are completely thread-safe.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+object wrapping that data.  For all other shareable types, a strict copy
+or proxy is made such that the corresponding objects continue to match.
+
+For now, shareable objects must be specifically supported internally
+by the Python runtime.
```
I expect sharing mutable data directly (without a proxy) will be uncommon  Regardless, shared mutable data ends up a matter of thread-safety, and that would have to be the responsibility of the user.  For `memoryview` (and buffers generally), I expect it to be manageable.

I'll make sure the PEP is clear about this.

gvanrossum replied:
Do you have locking primitives so users *can* serialize access to shared data if they want to go that way?

ericsnowcurrently replied:
Channels are basically the only locking primitive.  You pass a proverbial football around.

gvanrossum replied:
That's worth writing down.

ericsnowcurrently replied:
I'll add a short section about synchronization.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+
+...with one slight difference.  Rather than propagate the uncaught
+exception directly, we raise an ``interpreters.RunFailedError``
+with a snapshot of the uncaught exception (including its traceback)
```
I had planned on creating a lightweight custom class for the snapshot, but @iritkatriel pointed out to me that I could just use `traceback.TracebackException`.  That works pretty well.

ericsnowcurrently replied:
This became an option recently when @pablogsal cleaned up the default `sys.excepthook` to use `traceback.TracebackException`: gh-110702.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+...with one slight difference.  Rather than propagate the uncaught
+exception directly, we raise an ``interpreters.RunFailedError``
+with a snapshot of the uncaught exception (including its traceback)
+as the ``__cause__``.  Directly raising (a proxy of) the exception
```
If we make `traceback.TracebackException` inherit from `Exception` then this is a non-issue.  (The same goes if we use some other type for the snapshot.)

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,396 @@
+  it gets finalized.  Thus no interpreters created through
+  ``interpreters.create()`` will leak.
+
+* ``Interpreter.is_running()`` refers only to if there is a thread
+  running a script (code) in the interpreter's ``__main__`` module.
+  That basically means whether or not ``Interpreter.exec()`` is running
+  in some thread.  Code running in sub-threads is ignored.
```
The main practical value I see is being able to query before calling `interp.exec()`, to avoid getting an exception if it's already running.  Otherwise it's similar to `Thread.is_alive()` and would have similar utility.  I also imagine it could be helpful for people poking around and exploring how interpreters work.

ericsnowcurrently commented:
@gvanrossum, thanks for the review.  It's been supremely helpful.  I believe I've addressed all your comments.  FWIW, I'm fine with paring the text down, particularly in the introduction.  I just wanted a solid place to start from.

ericsnowcurrently commented:
Also, I'm aware of some formatting tweaks I still need to make to the API descriptions in the specification section.

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+-------------------------
+
+A Python process will have one or more threads running Python code
+(or otherwise interacting with the C-API).  Each of these threads
```
Re: https://docs.python.org/3/c-api/index.html

```suggestion
(or otherwise interacting with the C API).  Each of these threads
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+It is stored by the runtime in a thread-local variable, and may be
+looked up explicitly with ``PyThreadState_Get()``.  It gets set
+automatically for the initial ("main") thread and for
+``threading.Thread`` objects.  From the C-API it is set (and cleared)
+by ``PyThreadState_Swap()`` and may be set by ``PyGILState_Ensure()``.
+Most of the C-API requires that there be a current thread state,
```
```suggestion
``threading.Thread`` objects.  From the C API it is set (and cleared)
by ``PyThreadState_Swap()`` and may be set by ``PyGILState_Ensure()``.
Most of the C API requires that there be a current thread state,
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+
+As noted earlier, there is some runtime state that multiple threads
+share.  Some of it is exposed by the ``sys`` module, though much is
+used internally and not exposed explicitly or only through the C-API.
```
```suggestion
used internally and not exposed explicitly or only through the C API.
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+* low-level env vars
+* process memory (though allocators *are* isolated)
+* builtin types (e.g. dict, bytes)
+* singletons (e.g. None)
```
```suggestion
* singletons (e.g. ``None``)
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+      Returns ``other is self``.
+
+* ``empty()``
+      Return True if the queue is empty, False otherwise.
```
```suggestion
      Return ``True`` if the queue is empty, False otherwise.
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+      Return True if the queue is empty, False otherwise.
+
+* ``full()``
+      Return True if there are maxsize items in the queue.
+
+      If the queue was initialized with maxsize=0 (the default),
+      then full() never returns True.
```
```suggestion
      Return ``True`` if there are ``maxsize`` items in the queue.

      If the queue was initialized with ``maxsize=0`` (the default),
      then ``full()`` never returns ``True``.
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+      The object must be `shareable <Shareable Objects_>`_, which means
+      the object's data is passed through rather than the object itself.
+
+      If maxsize > 0 and the queue is full then this blocks until a free
+      slot is available.  If "timeout" is a positive number then it only
```
Italics for params: https://devguide.python.org/documentation/markup/#quick-reference

```suggestion
      If ``maxsize > 0`` and the queue is full then this blocks until a free
      slot is available.  If *timeout* is a positive number then it only
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+
+* ``get(timeout=None) -> object``
+      Pop the next object from the queue and return it.  Block while
+      the queue is empty.  If a positive timeout is provided and an
```
```suggestion
      the queue is empty.  If a positive *timeout* is provided and an
```

hugovk commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,798 @@
+   def worker():
+       interp = interpreters.create()
+       interp.prepare___main__(control=control, data=data)
+       interp.exec("""if True:
```
Looks like we're missing the end of the triple-quoted strings in these two examples?

It's more obvious from the preview:

https://pep-previews--3523.org.readthedocs.build/pep-0734/#synchronization

ericsnowcurrently commented:
@hugovk, thanks for the review.  I've addressed your comments.

gvanrossum reviewed (commented):
Much, much better! Here's my first batch of comments. I made it through the first paragraph of the Specification this time. I'll continue as I have time.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+Here are the relevant characteristics of the builtin ``exec()``,
+for comparison:
+
+* It runs in the current OS thread and pauses whatever was running there,
```
I keep getting confused about whether "it" refers to builtin.exec() or to Interpreter.exec() or both. I recommend being explicit and saying "Both ..." if that's what you mean (though some of the bullets seem to be calling out differences.)

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+  which resumes when ``exec()`` finishes.  No other threads are affected.
+  (To avoid pausing the current thread, run ``exec()``
+  in a ``threading.Thread``.)
+* It executes against a namespace, by default the ``__dict__`` of the
+  current module (i.e. ``globals()``).  It uses that namespace as-is
```
The default for builtins.exec() is not always globals() -- when run inside a function it is locals(). I think Interpreter.exec() always runs against the namespace of *that interpreter's* `__main__` module, right?

ericsnowcurrently replied:
Good catch!

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+* It executes against a namespace, by default the ``__dict__`` of the
+  current module (i.e. ``globals()``).  It uses that namespace as-is
+  and does not clear it before or after.
+* When code is run from the command-line (e.g. ``-m``) or the REPL,
+  the "current" module is always the ``__main__`` module
+  of the target (main) interpreter.
```
How is this relevant to Interpreter.exec()?

gvanrossum replied:
That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

ericsnowcurrently replied:
> That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

FWIW, I am moving this up to the Introduction and converting this specific bullet into a distinct, brief discussion of how the different CLI options relate to the relevant C API functions (e.g. `PyRun_String()`, `PyEval_EvalCode()`).  That's in the interest of identifying how `exec()`, `Interpreter.exec()`, and the CLI are related at a fundamental level.

> Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

Fair enough.  I'll clean that up.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+* When code is run from the command-line (e.g. ``-m``) or the REPL,
+  the "current" module is always the ``__main__`` module
+  of the target (main) interpreter.
+* It discards any object returned from the executed code::
```
Does this mean you can actually pass `func.__code__` to Interpreter.exec()? How does that work? If that's not supported, I'm confused -- why bring this up?

ericsnowcurrently replied:
I plan on supporting that right away, but didn't want to clutter the PEP with another dimension of usage.  I can remove this bullet if you think it's a distraction.

gvanrossum replied:
Have you implemented that yet? How were you planning to pass the code object around? Using marshal perhaps?

ericsnowcurrently replied:
Yeah, the implementation was surprisingly trivial.  I did indeed use marshal.

(FYI, I'm removing the bullet.)

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+  The exception is raised from the ``exec()`` call in the thread
+  that originally called ``exec()``.
+
+The only difference is that, rather than propagate the uncaught
```
This paragraph seems logically to be part of the last bullet. Why not indent it accordingly?

ericsnowcurrently replied:
I'm moving the paragraph elsewhere.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+Only objects that are specifically supported for passing
+between interpreters may be sent through a ``Queue``.  Note that the
+actual objects aren't sent, but rather their underlying data.
+However, the popped object will still be identical to the original.
```
"Identical" is probably not the term you're looking for here -- that would imply the two have the same address. I'd use equivalent or equal instead.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+
+* ``id``
+      (read-only) The target interpreter's ``InterpreterID`` object.
+      It is an int-like object with some internal bookkeeping.
```
Why does it have to be a special object? What does the extra bookkeeping do?

ericsnowcurrently replied:
Yeah, this doesn't need to be a special object.  I'll change it to `int`.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+      It is an int-like object with some internal bookkeeping.
+
+* ``maxsize``
+      Number of items allowed in the queue.
```
I'd add "Zero means unbounded."

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+      Return the hash of the queue's ``id``.  This is the same
+      as the hash of the ID's integer value.
+
+* ``__eq__(other)``
+      Return ``other is self``.
```
That's just inherited from `object`? Then it's not worth listing here. But maybe it should compare the `id` field?

ericsnowcurrently replied:
Good point.  It was previously comparing `.id` but I changed it to identity later.  I'll drop it.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+* ``__eq__(other)``
+      Return ``other is self``.
+
+* ``empty()``
+      Return ``True`` if the queue is empty, ``False`` otherwise.
+
+* ``full()``
+      Return ``True`` if there are ``maxsize`` items in the queue.
+
+      If the queue was initialized with ``maxsize=0`` (the default),
+      then ``full()`` never returns ``True``.
+
+* ``qsize()``
+      Return the number of items in the queue.
```
Maybe mention that these three are snapshots and their value can change any moment? E.g. qsize() may return 1 and yet get() could block -- if someone else snatched the item out of the queue before you.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+      ``interpreters.QueueFull``.  Otherwise is blocks forever.
+
+* ``put_nowait(obj)``
+      Like ``put()`` but effectively with an immediate timeout.
```
That differs from queue.Queue, which raises Full if the queue is full.

ericsnowcurrently replied:
That's what I was trying to communicate.  I'll clarify the text.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+      object hasn't been added to the queue in that many seconds
+      then raise ``interpreters.QueueEmpty``.
+
+* ``get_nowait([default]) -> object``
```
Any particular reason to add the default arg?

ericsnowcurrently replied:
It's a small thing but I personally found it useful.  If you think it's a distraction I could drop it.

gvanrossum replied:
I'd like as much API knowledge as possible to transfer between the different queue types we have (asyncio, threading, subinterpreters).

ericsnowcurrently replied:
You mean "keep the APIs as similar as possible"?  That's reasonable.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+to exactly match the corresponding object in the other interpreter.
+
+For some types (builtin singletons), the actual object is shared.
+For some, the object's underlying data is actually shared but each
+interpreter has a distinct object wrapping that data.  For all other
```
I guess that "wrapping" only happens for memoryview?

ericsnowcurrently replied:
yep

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+whereas for others only the underlying data (or even a copy or proxy)
+is shared.  Regardless, the guarantee of "shareable" objects is that
+corresponding objects in different interpreters will always strictly
+match each other.
```
I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-) I assume unicode normalization isn't going to be a problem.

ericsnowcurrently replied:
> I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-)

By "strictly match" I meant a deeper sort of equivalence than just equality.  The "shared" object would be indistinguishable from the original, whether it were the actual original object or a copy/proxy.  For oddballs like `NaN` or `inf` you'd get the right object on the other side (i.e.. `math.isnam()` and `math.isinf()`, respectively, would return true).  For mutable data, like buffers, mutation in one interpreter would be reflected in every other interpreter in which the object had been shared through a queue (or `prepare_main()`).

I'm definitely open to clearer phrasing.  In the meantime, I'll change it to "strictly equivalent".  Regardless, I'll try clarifying things with some additional text.

> I assume unicode normalization isn't going to be a problem.

Yeah, it shouldn't be a problem.

gvanrossum replied:
I think being explicit about what you just told me in the comment would be good. Especially the part about mutable buffers being (or acting as) truly shared. (Though all the C code manipulating these probably will need to use atomic operations to ensure the compiler actually writes to the buffer when you expect it to.)

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+(see `Synchronization`_), or by assigning sub-range exclusivity
+to individual interpreters.
+
+Regarding queues, the primary situation for sharing is where an
```
Maybe add "(described below)"?

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+-----------------------
+
+The new ``concurrent.futures.InterpreterPoolExecutor`` will be
+a subclass of ``concurrent.futures.ThreadPoolExecutor``, where each
```
Maybe just promise `c.f.Executor`? That's an abstract type that defines the interface.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+               results.put_nowait(res)
+           # Notify the results handler.
+           results.put_nowait(None)
+           """))
```
Extra ')'? Also in next example.

gvanrossum commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,800 @@
+                   res = capture_exception(exc)
+               results.put_nowait(res)
+           # Notify the results handler.
+           results.put_nowait(None)
```
Again, I'd just use `put()` everywhere.

gvanrossum reviewed (commented):
Here's the rest. This looks so much better than 554! Looking forward to more about the Executor (especially its failure modes). Also I think you don't describe exception handling from exec() yet?

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+1. create a new interpreter state
+2. create a new thread state
+3. set the thread state as current
+4. initialize the interpreter state using that thread state
```
That ordering is sort of an implementation detail.  Parts of interpreter init make use of the C-API, which needs a current thread state (e.g. for the current exception, imports).  It was simplest to just say we first set that tstate as current.

I'll add a brief note.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+"subinterpreters") since version 1.5 (1997).  The feature has been
+available via the :ref:`C API <python:sub-interpreter-support>`.
+
+Interpreters and Threads
```
It relates to how `Interpreter.exec()` and `Interpreter.prepare_main()` affect the current OS thread.  I suppose that is only sort of related to the proposed API.

gvanrossum replied:
But do they? Sure, `Interpreter.exec()` runs stuff in the current OS thread, but after it's done, everything is the same. I guess if you were looking from another OS thread at which interpreter or `ThreadState` is active in this OS thread (if there even is an API to do so) you'd see the other interpreter.

I don't see how `prepare_main()` affects the current OS thread -- does it? How?

ericsnowcurrently replied:
My primary concern is that people come in thinking that `Interpreter.exec()` automatically runs in a new thread, which isn't the case.  It's been a persistent confusion.  It's one of the reasons I renamed the method from `run()` to `exec()`.  That's why I've tried to spell out the relationship between interpreters and threads in several places, including this section.  If you aren't seeing the need to clarify relative to `Interpreter.exec()` then I've succeeded. 😄  If you think this section doesn't add much for understanding the execution context then I'd be glad to remove it.

> I don't see how `prepare_main()` affects the current OS thread -- does it? How?

`prepare_main()` must switch interpreters in current OS thread just like `exec()` does.  It has to modify the `__main__` module of the other interpreter, and the shared objects we're binding (at least the ones that aren't directly shareable) must get created in that interpreter.  For both cases we have to switch interpreters in the current OS thread.

gvanrossum replied:
Oh, I see, in order for `prepare_main()` to even update its interpreter's `__main__` it must make that interpreter the current interpreter. That's worth calling out in the spec for `prepare_main()`, but I don't worry about it the way I (still!) worry about `exec()`.

Regarding `exec()`, I think the problem is that running the code in the current OS thread is just not what people want. Given the purported use cases for subinterpreters, i.e. to utilize other cores, it's just basically never what you want, and the nice short name of the function makes it the first thing people look for.

Maybe it should just be renamed to `exec_in_current_os_thread()`? Or `exec_and_wait()`? Or `exec_synchronously()`, `exec_sync()`, or even `exec_here()`? Even if there will never be a function that spins up a new OS thread where the code is executed (though why not?), the longer name would alert everyone to the semantic pitfall. (And maybe direct them towards the interpreter pool API.)

ericsnowcurrently replied:
I see what you're saying.  I'll try `exec_sync()`.

I'll also consider adding a `run()` method as the variant that creates a thread to run in.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+To make an existing interpreter active in the current thread,
+we first make sure it has a thread state for the thread.  Then
+we call ``PyThreadState_Swap()`` like normal using that thread state.
+If the thread state for another interpreter was already current then
+it gets swapped out like normal and execution of that interpreter in
+the thread is thus effectively paused until it is swapped back in.
```
> Now I'm curious what happens if _no_ thread state was already current.

If there wasn't one then nothing's really that different.  The new interpreter get created and its new thread state gets set as current.

> Or if another thread state for the same interpreter was current.

That shouldn't be possible since we're just barely creating the new interpreter.  There wouldn't be any opportunity to create a second thread state for that interpreter.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+After initialization, the script or module or REPL is executed using
+them.  That execution happens in the interpreter's ``__main__`` module.
+
+When execution is done, the Python runtime is finalized in the main
+thread using the main interpreter.
```
> I'm guessing that "execution is done" is a way of saying "the main thread of the main interpreter exits", right?

right

> Does it do anything to alert other interpreters and threads?

Right away, finalization waits for all non-daemon threads to finish.  Later it finalizes all subinterpreters, which waits for each of their non-daemon threads to finish.  The various atexit funcs get called pretty early too.

ericsnowcurrently replied:
I'll add a note about this.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+The module defines the following functions:
+
+* ``get_current() -> Interpreter``
+      Returns an ``Interpreter`` object for the currently executing
+      interpreter.
```
I've pivoted to match `threading`, where there is only one object with a given interpreter ID.  That seems like it would be less surprising.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -0,0 +1,697 @@
+* ``create() -> Interpreter``
+      Create a new interpreter and return an ``Interpreter`` object
+      for it.  The interpreter will not be associated with any threads
+      until something is actually run in the interpreter.
```
Yeah, this is a bit unclear.  My intention was to reinforce the idea that an interpreter on its own isn't associated with any threads.  It isn't until the interpreter is used, e.g. `Interpreter.exec()`, that it gets tied to the current thread.  I'll make that more clear.

ericsnowcurrently commented on peps/pep-0734.rst:
```diff
@@ -511,23 +532,106 @@ Synchronization
+           while True:
+               token = control.get()
+               edit_data(data)
+               control.put_nowait(token)
```
good point

gvanrossum commented:
(Oh, I hope you're not using unresolved conversations as a TODO list. I've been marking things as resolved (adding a thumbs up) as soon as I'm happy with your response -- if you say "I'll add such-and-such" I haven't waited for such-and-such to actually appear in the text.)

ericsnowcurrently commented:
No worries.  I'm keeping track.

ericsnowcurrently commented:
@gvanrossum, I've made updates to address your review comments.  Thanks again for taking the time to provide such useful feedback. ❤️

I'm concerned that the PEP is getting a little too long, so let me know if there are any sections that you think are unnecessary or have too much detail, if you can spare me the time.

ericsnowcurrently commented:
FYI, once this PR is merged, I plan on updating the implementation (and update some benchmarks, to try it out) before posting the PEP for discussion.  That way I can tweak the PEP if something doesn't feel right in practice.

gvanrossum commented:
Are you awaiting more review from me? Im in Japan but can make a little time.

ericsnowcurrently commented:
Sweet of you to offer, but no.  If needed, we can chat when you are back.

ericsnowcurrently commented:
There are probably a handful of things to tweak, like trim back some of the fluff, but I'll do that in follow-up PRs.
